llvm/lib/Target/SPIRV/SPIRVBuiltins.cpp

```diff
@@ -622,6 +622,56 @@ static bool buildAtomicRMWInst(const SPIRV::IncomingCall *Call, unsigned Opcode,
   return true;
 }
 
+/// Helper function for building atomic flag instructions (e.g.
+/// OpAtomicFlagTestAndSet).
+static bool buildAtomicFlagInst(const SPIRV::IncomingCall *Call,
+                                unsigned Opcode, MachineIRBuilder &MIRBuilder,
+                                SPIRVGlobalRegistry *GR) {
+  const MachineRegisterInfo *MRI = MIRBuilder.getMRI();
+
+  Register PtrRegister = Call->Arguments[0];
+
+  Register MemSemanticsReg;
+  unsigned Semantics = SPIRV::MemorySemantics::SequentiallyConsistent;
+  if (Call->Arguments.size() >= 2) {
+    std::memory_order Order =
+        static_cast<std::memory_order>(getIConstVal(Call->Arguments[1], MRI));
+
+    Semantics =
+        getSPIRVMemSemantics(Order) |
+        getMemSemanticsForStorageClass(GR->getPointerStorageClass(PtrRegister));
+    if (Order == Semantics)
+      MemSemanticsReg = Call->Arguments[1];
+  }
+  if (!MemSemanticsReg.isValid())
+    MemSemanticsReg = buildConstantIntReg(Semantics, MIRBuilder, GR);
+
+  if (Opcode == SPIRV::OpAtomicFlagClear)
```

Comment:
Could we move this check `Opcode == SPIRV::OpAtomicFlagClear` to the assert condition?